### DISCUSSION TITLE
questions about loading mods

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1267] [Q&A] [answered]
I have a couple of questions about loading mods dynamically on an esp32, meaning over the network by an app that is running on the board. What I'd like to do is receive a mod over the network and save it to flash, then perform a reset such that this mod gets loaded. I'm trying to understand what is required to do that safely currently, and perhaps how it could be made more flexible. Questions:

1. I believe that before installing a new mod into flash a reset is required to ensure that no code/data from the old mod is in-use in any VM? And it appears that if one zeroes the first 16 bytes of the flash area this prevents the mod from being recognized & loaded after the reset?
2. After the mod is installed in flash it appears that only a reset can load it, there is no call to tell the VM to load it now? I.e. in total two resets are required, one before installing the mod and one after?

Looking at future options:

1. I gather that if the mod is in-use there is no safe way to "unload it" and overwrite the flash area because some code or data of the mod could be in-use somewhere in a VM? This means that either a reset prior to installing is required or an A/B two-partition scheme is required?
2. The main annoyance of a reset is the loss of connectivity, which can easily take 10-15 seconds to re-establish in the case of Wifi. Is there any existing machinery to perform a "soft-reset", e.g. stop all VMs and start the default VM from scratch? With some fiddling this could be done while keeping Wifi connected...
3. Can a mod be loaded, i.e. `fxMapArchive` be called after the VM is already running or does it have to happen right at creation time?
4. If I wanted to load two mods, would it "simply" be a matter of calling `fxMapArchive` twice, as is done in `xsHost.c:modInstallMods`? I'm not clear about the role of `kModulesStart` and wonder whether the memory mapping would result in a conflict.

**Comment by phoddie** [comment, marked as answer]
That's a lot of questions. I won't try answering all of them completely at once.

> I believe that before installing a new mod into flash a reset is required to ensure that no code/data from the old mod is in-use in any VM? And it appears that if one zeroes the first 16 bytes of the flash area this prevents the mod from being recognized & loaded after the reset?

Correct.

> After the mod is installed in flash it appears that only a reset can load it, there is no call to tell the VM to load it now? I.e. in total two resets are required, one before installing the mod and one after?

The default mod loader needs to run early because of symbol mapping.

>  I gather that if the mod is in-use there is no safe way to "unload it" and overwrite the flash area because some code or data of the mod could be in-use somewhere in a VM? This means that either a reset prior to installing is required or an A/B two-partition scheme is required?

Correct. If the mod is in-use, overwriting it is problematic since that would alter its bytecode and strings.

FWIW – in a project where flash was tight, instead of an A/B scheme, we have just a single mod partition. When an update is available, we invalidate the current mod, restart, download the mod over the previous one, and then restart. It is basically invisible to the user.

> The main annoyance of a reset is the loss of connectivity, which can easily take 10-15 seconds to re-establish in the case of Wifi. Is there any existing machinery to perform a "soft-reset", e.g. stop all VMs and start the default VM from scratch? With some fiddling this could be done while keeping Wifi connected...

The [mctest tool](https://www.moddable.com/documentation/tools/testing#testmc-app) -- what we use to run unit tests -- enables an obscure feature (see `MODDEF_XS_TEST`) that allows it to kill and relaunch the main VM without a restart of the MCU. We use the to be able to run tests back-to-back quickly (each test needs to run a fresh VM). It also has the side-effect you describe: keeping the Wi-Fi connection active across tests.

Finally, as I may have mentioned previously, XS has provisions for other mod loaders. We use that in some private projects but haven't yet published. It supports multiple mods active simultaneously, loading mods at any time, and running mods from RAM.  It works by building on the Compartment module loader in Hardened JavaScript. It works well, but you won't like it because it isn't well documented yet. ;)

**Reply by tve** [reply]
OK, I think this answers all the questions! Not sure why you think I wouldn't like the "other" mod loader: it's not like any of the questions I asked here came out of reading the docs... ;-) But I can proceed with the double-reset approach. Once I can show the results maybe you can be convinced to make the other mod loader available... :-) Thanks much!

**Reply by phoddie** [reply]
Sounds good. 

Note that the alternate mod loader doesn't fundamentally change the update behavior -- you still can't safely overwrite the archive of an active mod. That's no different than a firmware update or updating a native shared library or DLL (on a computer). Restarting after an update is common.

**Reply by tve** [reply]
> an obscure feature (see MODDEF_XS_TEST) that allows it to kill and relaunch the main VM without a restart of the MCU

**Update**: looks like it works now!

I've been unable to get this to work so far. I can call `fxAbort` such that `gThe` gets set to null. The problem is that doing this from a native module makes it happen during `modTimersExecute` and then stuff crashes and the CPU resets anyway. I wonder whether when used "properly" from debug, `gThe=NULL` happens during `modMessageService` and that's OK.

main.js:
```
import * as Sys from "sys"
import Timer from "timer"

trace("Hello world!\n")

Timer.delay(2000)
trace("Restarting...\n")
Sys.restart()
```

sys.js:
```
export function reset() @ "xs_reset"
export function restart() @ "xs_restart"
```

modSys.c:
```
#include "xs.h"

void xs_reset(xsMachine *the) {
#ifdef ESP32
	esp_restart();
#endif
}

#define XS_DEBUGGER_EXIT 0 // see xsAll.h
void xs_restart(xsMachine *the) {
	fxAbort(the, XS_DEBUGGER_EXIT);
}
```

manifest.json:
```
{
  "include": ["$(MODDABLE)/examples/manifest_base.json"],
  "modules": {
    "*": ["./*.js", "./*.c"]
  },
  "defines": {
    "XS_TEST": "1"
  }
}
```

**Reply by tve** [reply]
For the reset to work I had to make 2 changes:
- not use `Timer.delay(2000)` in my main.js and use `Timer.set` instead, as I should..
- in the "run the VM loop" in main.c add an if statement:
```
			modTimersExecute();
			if (gThe) modMessageService(gThe, modTimersNext());
			modInstrumentationAdjust(Turns, +1);
```

I notice that hitting the restart button is xsbug also restarts the VM so I'm wondering whether that's a better avenue...

**Reply by phoddie** [reply]
I think this is very close. The problem does appear to be triggering the soft-reset from a timer. It is a mistake to overload gThe to signal the soft-reset. I had good success by changing esp/xsPlatform.c > fxAbort to use gSoftReset:

```c
	if (XS_DEBUGGER_EXIT == status) {
		extern uint8_t gSoftReset;
		gSoftReset = 1;
		return;
	}
```

And modifying xsProj-espp32/main.c (if you are using a different CPU, change the corresponding main.c)
```c
#if MODDEF_XS_TEST
	uint8_t gSoftReset = 0;
#endif
```

and 

```c
#if MODDEF_XS_TEST
		xsMachine *the = gThe;
		while (!gSoftReset) {
			modTimersExecute();
			....
```

That gets your test working. 

If you are using the automatic connect to Wi-Fi feature, you will bump into a problem where it tries to connect to Wi-Fi when already connected. That causes a failure. To fix that, modify build/devices/esp32/setup/network.js to check for an existing connection:

```js
export default function (done) {

	if (!config.ssid) {
		trace("No Wi-Fi SSID\n");
		return done();
	}

	if (Net.get("SSID"))
		return done();
```

If this works out for you, I can look at integrating the changes.

**Reply by tve** [reply]
So... the trivial example works, but when I plug it into my app it crashes. What I have is in an HTTP handler: `Timer.set(() => Sys.restart(), 300)`. What I observe is that the VM loop exits and then it crashes in `fxDeleteMachine` in the `#ifndef mxLink` section where it seems to be calling a pile of destructors. I can't tell whether the issue is because one of the destructors is buggy (and normally never gets called) or whether the concept is problematic.

I notice that if I hit the restart button in xsbug the same crash happens. I need to figure out how to enable a esp-idf crash dump...

**Reply by phoddie** [reply]
A little more progress. The change to setup/network.js isn't necessary (you shouldn't use automatic Wi-Fi set-up if you want to maintain the connection across soft-reset).

The loop_task in main.c needs to reset gSoftReset or it will continuously reset. Oops.

```c
	while (true) {
#if MODDEF_XS_TEST
		gSoftReset = 0;
#endif
		gThe = modCloneMachine(NULL, NULL);

		modRunMachineSetup(gThe);

#if MODDEF_XS_TEST
		xsMachine *the = gThe;
		while (!gSoftReset) {
			modTimersExecute();
			...
```

With that, soft reset works for many cycles for me. That's consistent with the testmc use of the feature.

In general, the destructors have been tested in testmc. This is a little bit of an odd scenario so perhaps something unique is happening.

**Reply by phoddie** [reply]
FWIW – this is working reliably for me:

```js
import * as Sys from "sys"
import Timer from "timer"
import Net from "net";
import {Request} from "http"

trace("Hello world!\n")
trace(Net.get("SSID"), "\n");

let request = new Request({host: "www.example.com", path: "/", response: String});
request.callback = function(message, value, etc)
{
	if (Request.responseComplete == message) {
		trace(value);
		trace("\n");
	}
}
const timeout = Math.random() * 300;
trace("timeout " + timeout + "\n");
Timer.set(() => Sys.restart(), timeout);
```

The random timeout is to do the soft reset at different points during the HTTP request.

**Reply by tve** [reply]
Thanks! I will start bisecting...

**Reply by tve** [reply]
There seems to be a problem with the io network stack. Here's my main.js:
```js
import * as Sys from "sys"
import config from "mc/config"
import Timer from "timer"
import Preference from "preference"
import Net from "net"
import HTTPServer from "embedded:network/http/server"
import Listener from "embedded:io/socket/listener"
import WiFi from "wifi"

let ssid = Preference.get("wifi", "ssid")?.toString()
let password = Preference.get("wifi", "password")?.toString()
if (config.ssid) {
  Preference.set("wifi", "ssid", config.ssid.toString())
  ssid = config.ssid
}
if ("password" in config) {
  Preference.set("wifi", "password", config.password.toString())
  password = config.password
}

trace("Hello world!\n")
trace(`SSID: ${Net.get("SSID")}\n`)

if (Net.get("SSID") !== ssid) {
  new WiFi({ ssid, password }, (msg, code) => {
    if (msg == WiFi.gotIP) Timer.set(startHTTP, 200)
  })
} else {
  Timer.set(startHTTP, 200)
}

function startHTTP() {
  const server = new HTTPServer({
    io: Listener,
    port: 80,
    onConnect: connection => {
      trace(`HTTP connection!\n`)
    },
  })
  trace("Ready!\n")

  Timer.repeat(() => {
    trace("Alive...\n")
  }, 1000)

  Timer.set(() => {
    trace("Restarting...\n")
    Sys.restart()
  }, 2500)
}
```

If I comment out the `new HTTPServer` statement it works great, but with the statement it crashes.

When it works great, the xsbug output is:
```
No Wi-Fi SSID
Hello world!
SSID: undefined
Ready!
Alive...
Alive...
Restarting...
Soft reset!
No Wi-Fi SSID
Hello world!
SSID: tve-home
Ready!
Alive...
Alive...
Restarting...
Soft reset!
No Wi-Fi SSID
Hello world!
SSID: tve-home
Ready!
```
The second and third "SSID: tve-home" shows that it did do a soft reset. Run with the call to HTTPServer it always says `SSID: undefined`.
Looking at the serial2xsbug output, when it works right I see:
```
### fxRestart                                                                          
ets Jun  8 2016 00:22:57                                                               
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)                                
configsip: 0, SPIWP:0xee                                                               
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00                
mode:DIO, clock div:1                                                                  
load:0x3fff0030,len:344                                                                
load:0x40078000,len:14448                                                              
load:0x40080400,len:4                                                                  
ho 8 tail 4 room 4                                                                     
load:0x40080404,len:3268                                                               
entry 0x40080580                                                                       
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
### set time                                                                           
```
(I have MX_TRACE_COMMANDS set to 1)
and with the HTTPServer call I get:
```
### fxRestart                                                                          
ets Jun  8 2016 00:22:57                                                               
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)                                
configsip: 0, SPIWP:0xee                                                               
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00                
mode:DIO, clock div:1                                                                  
load:0x3fff0030,len:344                                                                
load:0x40078000,len:14448                                                              
load:0x40080400,len:4                                                                  
ho 8 tail 4 room 4                                                                     
load:0x40080404,len:3268                                                               
entry 0x40080580                                                                       
### set time                                                                           
ets Jun  8 2016 00:22:57                                                               
rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)                                 
configsip: 0, SPIWP:0xee                                                               
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1                                                                  
load:0x3fff0030,len:344                    
load:0x40078000,len:14448                                                              
load:0x40080400,len:4                                                                  
ho 8 tail 4 room 4                                                                     
load:0x40080404,len:3268                                                               
entry 0x40080580                                                                       
### set time                                                                           
ets Jun  8 2016 00:22:57                                                               
rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)                                 
...
```

For reference, my manifest.json:
```
{
  "include": [
    "$(MODDABLE)/examples/manifest_base.json",
    "$(MODDABLE)/examples/manifest_net.json",
    "$(MODULES)/files/preference/manifest.json",
    "$(MODDABLE)/examples/io/listener/httpserver/manifest_httpserver.json"
  ],
  "modules": {
    "*": ["./*.js", "./*.c"]
  },
  "defines": {
    "XS_TEST": "1"
  }
}
```

**Reply by phoddie** [reply]
An I correct in assuming you aren't actually establishing a connection to the HTTP server in these tests run? So, the only network instance that is active is the Listener socket.

**Reply by tve** [reply]
Correct.

**Reply by tve** [reply]
Here's the backtrace when it crashes in fxDeleteMachine:
```
Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.     
EXCVADDR: 0x0201014f
PC: 0x4012762d: tcp_recv at /home/sw/moddable/esp-idf5/components/lwip/lwip/src/core/tcp.c:2026
BT-0: 0x4012762a: tcp_recv at /home/sw/moddable/esp-idf5/components/lwip/lwip/src/core/tcp.c:2023
BT-1: 0x401035d5: closeSocket at /home/sw/moddable/mod2/modules/network/socket/lwip/modSocket.c:421
BT-2: 0x40104396: xs_socket_destructor at /home/sw/moddable/mod2/modules/network/socket/lwip/modSocket.c:448
 (inlined by) xs_socket_destructor at /home/sw/moddable/mod2/modules/network/socket/lwip/modSocket.c:441
BT-3: 0x40104e92: xs_listener_destructor at /home/sw/moddable/mod2/modules/network/socket/lwip/modSocket.c:1179 (discriminator 3)
BT-4: 0x400d63af: fxDeleteMachine at /home/sw/moddable/mod2/xs/sources/xsAPI.c:1590
BT-5: 0x400d48b1: loop_task at /home/sw/moddable/mod2/build/tmp/esp32/nodemcu/debug/repro-restart/xsProj-esp32/main/main.c:148
```

**Reply by phoddie** [reply]
Are you sure that's the backtrace for the script above? The backtrace shows the original socket but the script uses ECMA-419 socket/listener.

**Reply by phoddie** [reply]
That is the backtrace. The mistake is here:

https://github.com/Moddable-OpenSource/moddable/blob/0bd542edecf011234826b52c8be1504672b95dc9/modules/io/socket/lwip/tcp.c#L638-L641

Change `xs_listener_destructor` to `xs_listener_destructor_` and all works well.

**Reply by tve** [reply]
Confirmed! Thanks for all the help & fixes!

Can you consider supporting something like `fxAbort(the, XS_RESTART_EXIT)` in all build types and not being tied to XS_TEST?

**Reply by phoddie** [reply]
Glad that's working. I think soft reset can be separated from test. This would be for the ESP32 family to start.

I don't think it needs to go through fxAbort. That's an XS function and this is a runtime behavior. Maybe just something like `modSoftReset()`.

**Reply by phoddie** [reply]
I've implemented `modSoftReset()` for all ESP32 family hosts. Soft reset is enabled by setting "softreset" to `1` in the `defines` section of the manifest. That is now set by `test262` and `testmc` in addition to `XS_TEST`. Your http server example above works with minor changes to define `softreset` and call `modSoftReset()`. We'll get that pushed here later today so you can verify.

**Reply by tve** [reply]
Thanks!! Is there interest in a Sys module as I've started? If so, where should it be placed/called? (I'm also happy to just keep it in my projects.)

**Reply by phoddie** [reply]
> Is there interest in a Sys module as I've started? 

Yes, but... it kind of overlaps with the `System` module placeholder in ECMA-419 that I'd like to remove. ;) There are a handful of these little things (restart) that are trivial but need a home. System is always a dangerous term. Maybe Utilities is better?

**Reply by tve** [reply]
Micropython uses Machine, which works pretty well...

**Reply by phoddie** [reply]
That's a little hardware-ish sounding and also kind of overlaps with "virtual machine", both of which could be confusing. We already have `device` -- I'm not sure we can easily explain when to use `device` versus `machine`?

**Reply by tve** [reply]
`device` broke down for me when I started seeing all the `device.network.*` stuff...
I believe System is pretty good. Another option I can come up with is Power or LowPower: that could include functions to control power states, deep sleep, etc, and reset/shutdown are not a huge stretch from there... Yet another name might be Processor. Processor.reset, Processor.softReset, Processor.deepSleep, Processor.clockRate, ... seem to work pretty well.

**Reply by phoddie** [reply]
I prefer Processor to Power. Processor might not scale well to multi-core silicon (not so rare these days) though. I'm thinking of cases where a setting might apply to one core but not another. Maybe I'm over thinking it though.

All that kind of thing varies quite a bit by silicon and RTOS. An alternative could be just to have an "ESP32" object that provides these things in a way that makes sense for ESP32 family. That sidesteps the issue of trying to find APIs that work across silicon vendors and then explaining how they map to the underlying silicon. Or that could hang off Processor as in `Processor.ESP32` so that common things (restart, at least) are consistent across silicon and vendor specific things are explicitly accessed as such.

**Reply by tve** [reply]
Processor sounds really good to me. Processor.ESP32 for family-specific stuff also. For multicore I don't see a problem in passing the core number/designator to relevant functions (could be optional/defaulted param), but I suspect that's going to be silicon specific almost regardless.

**Reply by phoddie** [reply]
> For multicore I don't see a problem in passing the core number/designator to relevant functions (could be optional/defaulted param), but I suspect that's going to be silicon specific almost regardless.

Sure.
